src/stream/src/common/table/state_table.rs

```diff
@@ -938,6 +934,21 @@ where
             .map(get_second))
     }
 
+    pub async fn iter_all_vnodes_with_pk_range(
```

Comment:
It seems we can move the `merge_sort` function into this function. It could be more friendly for users.

Reply:
I previously tried it but had some difficulty, due to the following error:
```sh
error[E0391]: cycle detected when type-checking `common::table::state_table::<impl at src/stream/src/common/table/state_table.rs:875:1: 875:41>::iter_all_vnodes_with_pk_range`
    --> src/stream/src/common/table/state_table.rs:938:5
     |
938  | /     pub async fn iter_all_vnodes_with_pk_range(
939  | |         &self,
940  | |         pk_range: &(Bound<impl Row>, Bound<impl Row>),
941  | |         prefetch_options: PrefetchOptions,
942  | |     ) -> StreamExecutorResult<impl Stream<Item = StreamExecutorResult<(Bytes, OwnedRow)>>> {
     | |__________________________________________________________________________________________^
     |
note: ...which requires computing type of `common::table::state_table::RowStreamWithPk::{opaque#0}`...
    --> src/stream/src/common/table/state_table.rs:1102:5
     |
1102 |     impl Stream<Item = StreamExecutorResult<(Bytes, OwnedRow)>> + 'a;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: ...which again requires type-checking `common::table::state_table::<impl at src/stream/src/common/table/state_table.rs:875:1: 875:41>::iter_all_vnodes_with_pk_range`, completing the cycle
note: cycle used when computing type of `common::table::state_table::RowStream::{opaque#0}`
    --> src/stream/src/common/table/state_table.rs:1100:5
     |
1100 |     impl Stream<Item = StreamExecutorResult<OwnedRow>> + 'a;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Here's the updated definition:
```rs

    pub async fn iter_all_vnodes_with_pk_range(
        &self,
        pk_range: &(Bound<impl Row>, Bound<impl Row>),
        prefetch_options: PrefetchOptions,
    ) -> StreamExecutorResult<impl Stream<Item = StreamExecutorResult<(Bytes, OwnedRow)>>> {
        let mut vec = Vec::with_capacity(self.vnodes.count_ones());
        for vnode in self.vnodes.iter_vnodes() {
            vec.push(
                self.iter_key_and_val_with_pk_range(pk_range, vnode, prefetch_options)
                    .await?,
            )
        }
        let pinned_iter: Vec<_> = vec.into_iter().map(Box::pin).collect_vec();
        let iter = merge_sort(pinned_iter);
        merge_sort(vec);
        Ok(iter)
    }
```

Wonder if any rust expert have advice.. @BugenZhao

Reply:
The following codes could be compiled.
```rust
    pub async fn iter_all_vnodes_with_pk_range(
        &self,
        pk_range: &(Bound<impl Row>, Bound<impl Row>),
        prefetch_options: PrefetchOptions,
    ) -> StreamExecutorResult<RowStreamWithPk<'_, S, SD>> {
        let mut vec = Vec::with_capacity(self.vnodes.count_ones());
        for vnode in self.vnodes.iter_vnodes() {
            vec.push(
                self.iter_key_and_val_with_pk_range(pk_range, vnode, prefetch_options)
                    .await?,
            )
        }
        let pinned_iter: Vec<_> = vec.into_iter().map(Box::pin).collect_vec();
        let iter = merge_sort(pinned_iter);
        Ok(iter)
    }


```

```rust
#[try_stream(ok=(Bytes, OwnedRow), error=StreamExecutorError)]
pub async fn merge_sort<R>(streams: Vec<R>)
where
    R: Stream<Item = StreamExecutorResult<(Bytes, OwnedRow)>> + Unpin,
{
    let mut heap = BinaryHeap::new();
    for mut stream in streams {
        if let Some(peeked) = stream.next().await.transpose()? {
            heap.push(Node { stream, peeked });
        }
    }
    while let Some(mut node) = heap.peek_mut() {
        // Note: If the `next` returns `Err`, we'll fail to yield the previous item.
        yield match node.stream.next().await.transpose()? {
            // There still remains data in the stream, take and update the peeked value.
            Some(new_peeked) => std::mem::replace(&mut node.peeked, new_peeked),
            // This stream is exhausted, remove it from the heap.
            None => PeekMut::pop(node).peeked,
        };
    }
}

```